Support decoding optional headers in the OTLP-Arrow receiver

jmacd opened the pull request:
**Description:** 

This sets the incoming gRPC headers from the new optional bytes field added here: https://github.com/f5/otel-arrow-adapter/pull/85. The behavior of this change matches #35 in the exporter.

The gRPC stream's incoming context is merged with the per-request metadata calculated for the individual data item. This addresses a problem described in greater detail in https://github.com/open-telemetry/opentelemetry-collector/issues/6965. There is a suggestion that OTC would be better off with a new extension type for non-auth-related metadata, but for now the Arrow receiver expects the exporter to run the Auth extension on the request context and serialize the result; the best the receiver can do is merge the two metadatas.

Part of #33

lquerel reviewed (approved):
LGTM

jmacd commented:
Updated the receiver to use `client.Info.Metadata` which is the Collector's standard mechanism for receivers to convey headers. PTAL especially @paivagustavo, thanks.

kristinapathak commented on receiver/otlpreceiver/config.go:
```diff
@@ -40,6 +40,9 @@ type Protocols struct {
 type ArrowSettings struct {
 	Enabled bool `mapstructure:"enabled"`
+
+	// TODO: Do we want a flag to disable the propagation of
```
I don't think we need a flag.

jmacd replied:
Fixed!

jmacd commented:
Note: I've updated this PR with a slight fix. This code now respects the GRPCServerSettings.IncludeMetadata flag. This must be set for headers to be set into the client context.

jmacd commented:
Due to the changes in #35 I will now update this PR to merge the incoming context (from the stream, including the static metadata and any the Auth plugin returns for a background context) with the per-request metadata that is transmitted per batch by the exporter.

jmacd commented:
This PR now matches #35: when `IncludeMetadata` is set the stream context's gRPC incoming metadata is merged with the per-request context. this means the context constructed by the OTLP-Arrow receiver contains the results of evaluating the Auth extension twice, once on the background context and once on the per-request context. PTAL.

jmacd commented:
Did some refactoring of the new header-decoding logic to make it more readable, added a few tests and corrected lint errors. [88f4210](https://github.com/open-telemetry/experimental-arrow-collector/pull/34/commits/88f42107262b95602da805f8d147e335653e6c41)
